### PR TITLE
Add access to inputs and outputs of op overloads

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72729

### CODE REVIEW

File: torch/_ops.py [hunk 1]
```diff
@@ -43,27 +44,34 @@ def __str__(self):
...
-        return getattr(self._op, key)
-
     # `my_namespace::my_op`
     @property
     def name(self):
         return "{}.{}".format(*self._schema.name.split("::"))
 
     @property
-    def overload_name(self):
+    def overloadname(self):
```
Comment (albanD) [inline torch/_ops.py:53]: We should land on stable name for these. As this is going to become public API.

File: torch/csrc/jit/python/init.cpp [hunk 2]
```diff
@@ -1468,16 +1468,50 @@ void initJITBindings(PyObject* module) {
     return type.value();
   });
 
-  py::class_<FunctionSchema>(m, "FunctionSchema")
+    py::class_<Argument>(m, "Argument")
```
Comment (albanD) [inline torch/csrc/jit/python/init.cpp:1471]: Why is this moved up? It makes it hard to see what was added.

File: torch/csrc/jit/python/init.cpp [hunk 3]
```diff
@@ -1468,16 +1468,50 @@ void initJITBindings(PyObject* module) {
...
+          })
+      .def(
+          "has_default_value",
+          [](Argument& self) -> py::bool_ {
+            return self.default_value().has_value();
+          })
+      .def_property_readonly("kwarg_only", [](Argument& self) -> bool {
+        return self.kwarg_only();})
+      .def_property_readonly(
+          "is_mutable", [](Argument& self) -> bool { return self.is_out(); });
```
Comment (albanD) [inline torch/csrc/jit/python/init.cpp:1504]: What does the "is_out" means? We have some inplace ops that are not out= variants.

File: torch/csrc/jit/python/init.cpp [hunk 4]
```diff
@@ -1468,16 +1468,50 @@ void initJITBindings(PyObject* module) {
...
+    py::class_<FunctionSchema>(m, "FunctionSchema")
       .def_property_readonly(
           "name", [](FunctionSchema& self) { return self.name(); })
       .def_property_readonly(
           "overload_name",
           [](FunctionSchema& self) { return self.overload_name(); })
       .def_property_readonly(
-          "arguments", [](FunctionSchema& self) { return self.arguments(); })
+          "arguments", [](FunctionSchema& self) { return py::list(py::cast(self.arguments())); })
```
Comment (albanD) [inline torch/csrc/jit/python/init.cpp:1512]: Why was this not needed before?

File: torch/csrc/jit/python/init.cpp [hunk 5]
```diff
@@ -1468,16 +1468,50 @@ void initJITBindings(PyObject* module) {
...
+            }
+            IValue v = *self.default_value();
+            return toPyObject(std::move(v));
+          })
+      .def(
+          "has_default_value",
+          [](Argument& self) -> py::bool_ {
+            return self.default_value().has_value();
+          })
+      .def_property_readonly("kwarg_only", [](Argument& self) -> bool {
```
Comment (albanD) [inline torch/csrc/jit/python/init.cpp:1501]: nit: formatting is quite inconsistent

File: test/test_per_overload_api.py [hunk 6]
```diff
@@ -60,5 +62,16 @@ def test_basics_opoverload(self):
...
 
+        add_outoverload = add_packet.out
+        self.assertEqual(str(add_outoverload.inputs), "[Tensor self, Tensor other, Scalar alpha, Tensor& out]")
+        self.assertTrue(add_outoverload.inputs[3].is_mutable)
+        self.assertTrue(not add_outoverload.inputs[2].is_mutable)
+        self.assertTrue(add_outoverload.inputs[3].kwarg_only)
+        self.assertTrue(add_outoverload.inputs[2].has_default_value)
+        self.assertEqual(add_outoverload.inputs[2].default_value, 1)
+        self.assertEqual(add_outoverload.inputs[2].name, "alpha")
+        self.assertEqual(add_outoverload.inputs[2].type, torch._C.NumberType.get())
```
Comment (anjali411) [inline test/test_per_overload_api.py:73]: ideally we wouldn't want these Type class objects to have their module set to `torch` instead of `torch._C` but changing this would be bc breaking for people working with TorchScript graphs (these types are being used their -- see `test_jit.py`)

File: torch/csrc/jit/python/init.cpp [hunk 7]
```diff
@@ -1521,14 +1533,15 @@ void initJITBindings(PyObject* module) {
...
+      .def_property_readonly(
           "has_default_value",
           [](Argument& self) -> py::bool_ {
             return self.default_value().has_value();
           })
       .def_property_readonly("kwarg_only", [](Argument& self) -> bool {
-        return self.kwarg_only();
-      });
+        return self.kwarg_only();})
+      .def_property_readonly(
```
Comment (anjali411) [inline torch/csrc/jit/python/init.cpp:1543]: add similar info for views

File: torch/csrc/jit/python/init.cpp [hunk 8]
```diff
@@ -1504,7 +1504,19 @@ void initJITBindings(PyObject* module) {
           })
       .def_property_readonly(
           "is_mutable", [](FunctionSchema& self) { return self.is_mutable(); });
+
   py::class_<Argument>(m, "Argument")
+      .def("__repr__", [](Argument& self) {
+        std::ostringstream s;
+        if (self.alias_info() && !self.alias_info()->isWrite()) {
```
Comment (anjali411) [inline torch/csrc/jit/python/init.cpp:1511]: Alban's suggestion:`
we should have a more pythonic print for Argument e.g.,`Argument(cpp_type='const Tensor&', name='self')`

maybe we should overload inputs print

Comment (anjali411) [reply torch/csrc/jit/python/init.cpp:1511]: @ezyang what do you suggest?

**Comment by ezyang** [comment]
Would be nice to have a PR description

File: test/test_per_overload_api.py [hunk 9]
```diff
@@ -32,6 +32,8 @@ def test_basics_opoverloadpacket(self):
         self.assertRaises(AttributeError, lambda: add_packet.foo)
 
     def test_basics_opoverload(self):
+        # add is ony used as an example here. It is ok to update the test
```
Comment (ezyang) [inline test/test_per_overload_api.py:35]: only

File: torch/csrc/jit/python/init.cpp [hunk 10]
```diff
@@ -1349,14 +1361,15 @@ void initJITBindings(PyObject* module) {
...
           "has_default_value",
           [](Argument& self) -> py::bool_ {
             return self.default_value().has_value();
           })
       .def_property_readonly("kwarg_only", [](Argument& self) -> bool {
-        return self.kwarg_only();
-      });
+        return self.kwarg_only();})
+      .def_property_readonly(
+          "is_mutable", [](Argument& self) -> bool { return self.alias_info() && self.alias_info()->isWrite(); });
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:1372]: If we add a method here can we mirror it in the C++ API? Let's keep the APIs lined up as much as possible.

File: torch/csrc/jit/python/init.cpp (hunk 8 above)
Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:1339]: IDK, since you can't construct arguments by hand I don't think I care too much. I'd probably recommend lining up the C++ printing and the Python printing. There is an `operator<<` defined on Argument, you should use that.

File: test/test_per_overload_api.py [hunk 11]
```diff
@@ -60,5 +62,16 @@ def test_basics_opoverload(self):
...
+        add_outoverload = add_packet.out
+        self.assertEqual(str(add_outoverload.inputs), "(Tensor self, Tensor other, Scalar alpha, Tensor& out)")
+        self.assertTrue(add_outoverload.inputs[3].is_mutable)
+        self.assertTrue(not add_outoverload.inputs[2].is_mutable)
+        self.assertTrue(add_outoverload.inputs[3].kwarg_only)
+        self.assertTrue(add_outoverload.inputs[2].has_default_value)
+        self.assertEqual(add_outoverload.inputs[2].default_value, 1)
+        self.assertEqual(add_outoverload.inputs[2].name, "alpha")
+        self.assertEqual(add_outoverload.inputs[2].type, torch.NumberType.get())
+        self.assertEqual(add_outoverload.inputs[0].type, torch.TensorType.get())
```
Comment (ezyang) [inline test/test_per_overload_api.py:74]: So... are we going to document the Argument API?

**Comment by ezyang** [comment]
I think we need to be a little more intentional about the API here. You are reusing the existing bindings for JIT schema Argument, which is a good idea, but as far as I can tell these bindings are completely undocumented. If we are serious about wanting people to use them they need to be documented. And as we are documenting, we may notice things in the bindings that don't make sense; we should flag those for future refactoring.

File: torch/_ops.py [hunk 12]
```diff
@@ -65,7 +64,17 @@ def overloadpacket(self):
     def op(self):
         return self._op
 
-    # TODO: add more methods to expose information about input and output arguments
+    # returns a list of torch._C.Argument class objects
+    # users can query the following on each Argument object:
+    # name, type, N, kwarg_only, has_default_value, default_value, is_mutable
+    @property
+    def inputs(self):
+        return tuple(self._schema.arguments)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_ops.py:72]: This feels like a binding problem then: why were they bound as lists rather than tuples...

File: torch/csrc/jit/python/init.cpp (hunk 10 above)
Comment (anjali411) [reply torch/csrc/jit/python/init.cpp:1372]: sounds good
